binderhub/launcher.py

```diff
@@ -106,60 +117,67 @@ def username_from_repo(self, repo_url):
         # add a random suffix to avoid collisions for users on the same image
         return '{}-{}'.format(prefix, ''.join(random.choices(SUFFIX_CHARS, k=SUFFIX_LENGTH)))
 
-    async def launch(self, image, username):
+    async def launch(self, image, username, server_name='', repo_url=''):
         """Launch a server for a given image
 
-        - creates the user on the Hub
-        - spawns a server for that user
+        - creates a temporary user on the Hub if authentication is not enabled
+        - spawns a server for temporary/authenticated user
         - generates a token
         - returns a dict containing:
           - `url`: the URL of the server
+          - `image`: image spec
+          - `repo_url`: the url of the repo
           - `token`: the token for the server
         """
         # TODO: validate the image argument?
 
-        # create a new user
-        app_log.info("Creating user %s for image %s", username, image)
-        try:
-            await self.api_request('users/%s' % username, body=b'', method='POST')
-        except HTTPError as e:
-            if e.response:
-                body = e.response.body
-            else:
-                body = ''
-            app_log.error("Error creating user %s: %s\n%s",
-                username, e, body,
-            )
-            raise web.HTTPError(500, "Failed to create temporary user for %s" % image)
-
-        # generate a token
-        token = base64.urlsafe_b64encode(uuid.uuid4().bytes).decode('ascii').rstrip('=\n')
+        if self.create_user:
+            # create a new user
+            app_log.info("Creating user %s for image %s", username, image)
+            try:
+                await self.api_request('users/%s' % username, body=b'', method='POST')
+            except HTTPError as e:
+                if e.response:
+                    body = e.response.body
+                else:
+                    body = ''
+                app_log.error("Error creating user %s: %s\n%s",
+                    username, e, body,
+                )
+                raise web.HTTPError(500, "Failed to create temporary user for %s" % image)
+        elif server_name == '':
```

Comment:
`elif server_name == '':` -> `elif server_name:` should be enough to check if a string is not empty.